### PR TITLE
gtkcord4: init at 0.0.6

### LINKED ISSUE #96936: package request: cchat-gtk/cchat-discord: a chat client with discord backend
**Project description**
- cchat-gtk appears to be a generic chat frontend
- cchat-discord appears to be a discord-capable backend

gtkcord3 was the predecessor and was quite nice, I'm excited to get this packaged.

**Metadata**

* homepage URL: https://github.com/diamondburned/gtkcord3
* source URL: https://github.com/diamondburned/gtkcord3
* license: gpl3
* platforms: linux

### PR DESCRIPTION
**Pull request by urandom2**
Fixes #96936

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/gtkcord4/about.patch [hunk 1]
```diff
@@ -0,0 +1,9 @@
+diff --git a/internal/gtkcord/window/about/about.go b/internal/gtkcord/window/about/about.go
+deleted file mode 100644
+index 6767b1e..0000000
+--- b/internal/gtkcord/window/about/about.go
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/gtkcord4/about.patch:4]: We can just rm this file in postPatch

File: pkgs/applications/audio/gtkcord4/default.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+, gdk-pixbuf
+, glib
+, gobject-introspection
+, graphene
+, gtk4
+, lib
+, pango
+, pkg-config
+}:
+buildGoModule rec {
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/gtkcord4/default.nix:13]:
```suggestion
}:

buildGoModule rec {
```

File: pkgs/applications/audio/gtkcord4/default.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    cairo
+    gdk-pixbuf
+    glib
+    gobject-introspection
+    graphene
+    gtk4
+    pango
+  ];
+
+  nativeBuildInputs = [ pkg-config ];
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/gtkcord4/default.nix:36]: Normally we put nativeBuildInputs before buildInputs.

File: pkgs/applications/audio/gtkcord4/default.nix [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    rm internal/gtkcord/window/about/about.go
+  '';
+
+  nativeBuildInputs = [ pkg-config ];
+
+  buildInputs = [
+    cairo
+    gdk-pixbuf
+    glib
+    gobject-introspection
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/gtkcord4/default.nix:35]: You are sure about this? Normally gobject-introspection goes into nativeBuildInputs.

Comment (urandom2) [reply pkgs/applications/audio/gtkcord4/default.nix:35]: I am more than happy to move it, but it does compile like this; maybe go only links against the shared library and does not call the binary?

**Comment by gador** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 200179` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>gtkcord4</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
gtkcord4:
<details>

got build log for '/nix/store/k1q8zkd9ni485fwwh2rvy382kmcdnkq3-gtkcord4-0.0.5' from 'daemon'
When evaluating attribute ‘gtkcord4’:
warning: unclear-gpl
`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/gtkcord4/default.nix:48:5:

```
   |
48 |     license = licenses.gpl3;
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
</details>
</details>

File: pkgs/applications/audio/gtkcord4/default.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  src = fetchFromGitHub {
+    owner = "diamondburned";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-AjHqKIW3pfc6AGDUMBbQ8FUK8ciUxngM45gutuSeTNM=";
+  };
+
+  postPatch = ''
+    rm internal/gtkcord/window/about/about.go
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/gtkcord4/default.nix:26]: Please add a short comment why we remove the file

File: pkgs/applications/audio/gtkcord4/default.nix [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    glib
+    graphene
+    gtk4
+    pango
+  ];
+
+  nativeBuildInputs = [
+    gobject-introspection
+    pkg-config
+  ];
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/gtkcord4/default.nix:41]: Normally we would swap position with buildInputs

**Review by hmenke — CHANGES_REQUESTED** [review]

**Review by hmenke — APPROVED** [review]
Works. The login token can be extracted from the regular Discord desktop application's local storage database:

```
grep -aoP '[\w-]{24}\.[\w-]{6}\.[\w-]*' ~/.config/discord/Local\ Storage/leveldb/*.ldb
```

Screenshot from 2022-11-30 11-25-44

**Comment by hmenke** [comment]
Just wanted to upload a file which crashed the application.
```
2022/12/02 19:47:35 Error: GLib-GIO: Settings schema 'org.gtk.gtk4.Settings.FileChooser' is not installed
SIGTRAP: trace trap
```
Looks like some sort of linkage to schemas is missing.

**Comment by urandom2** [comment]
I do not see this on my system, is it possible there is a runtime dependency we need to install or wrap?

**Comment by hmenke** [comment]
I think we need to add `wrapGAppsHook` to `nativeBuildInputs`. Building right now, but the build takes quite a long time and doesn't seem to be parallelized.

**Comment by hmenke** [comment]
But the build cannot be parallelized. That's a limitation of cgo: https://github.com/diamondburned/gtkcord4/issues/78

**Comment by SuperSandro2000** [comment]
Please create a PR with that and make sure there is no doubel wrapping happening

**Review by hmenke — APPROVED** [review]

**Comment by hmenke** [comment]
There is also an issue with finding GStreamer. I just tried to play back a video which crashed the application.
```
2022/12/13 12:26:53 got unknown image type "video/mp4", continuing anyway
2022/12/13 12:26:53 Error: GStreamer-Play: GstPlay: 'playbin' element not found, please check your setup
Trace/breakpoint trap
```

**Comment by urandom2** [comment]
that seems fine by me, I am not terribly familiar with GStreamer; are these optional dependencies, or pretty core to your usage of gktcord?

**Pull request merged**